relay-server/src/actors/upstream.rs

```diff
@@ -246,21 +325,71 @@ impl UpstreamRelay {
             //   1. Have own connection pool for `store` requests
             //   2. Buffer up/queue/synchronize events before creating the request
             .wait_timeout(self.config.event_buffer_expiry())
-            .conn_timeout(self.config.http_connection_timeout())
             // This is the timeout after wait + connect.
             .timeout(self.config.http_timeout())
+            .conn_timeout(self.config.http_connection_timeout())
+            .to_tracked(self.http_request_finished_notifier.clone())
             .map_err(UpstreamRequestError::SendFailed)
-            .and_then(handle_response);
+            .and_then(handle_response)
+            .then(|x| {
+                response_sender.send(x).ok();
+                futures::future::ok(())
+            });
+        Box::new(future)
+    }
+
+    fn send_request<P, F>(
+        &mut self,
+        priority: RequestPriority,
+        method: Method,
+        path: P,
+        build: F,
+    ) -> ResponseFuture<ClientResponse, UpstreamRequestError>
+    where
+        F: 'static + Send + FnOnce(&mut ClientRequestBuilder) -> Result<ClientRequest, ActixError>,
+        P: AsRef<str>,
+    {
+        let (tx, rx) = oneshot::channel::<Result<ClientResponse, UpstreamRequestError>>();
+
+        let request = UpstreamRequest {
+            method,
+            path: path.as_ref().to_owned(),
+            response_sender: tx,
+            build: Box::new(build),
+        };
+        match priority {
+            RequestPriority::Low => {
+                self.lp_messages.push_front(request);
+                metric!(
+                    histogram(RelayHistograms::UpstreamLowPriorityMessageQueueSize) =
+                        self.lp_messages.len() as u64
```

Comment:
Please use a single metric here and tag the queue priority. This will allow you to deduplicate the code like this (pseudo-code):

```rust
let queue = match priority {
  RequestPriority::Low => &mut self.lp_messages,
  RequestPriority::High => &mut self.hp_messages,
};

queue.push_front(request);
metric!(histogram(RelayHistograms::UpstreamQueueSize = ..., priority = priority));
```

Reply:
done, didn't manage to use directly `priority` in the histogram, I'm using a `fn name(&self)` method i.e. `priority.name()`